### ISSUE TITLE
Add POSIXlt method for IDate, and use it for as.IDate.POSIXct

### ISSUE THREAD

**Issue by MichaelChirico** [Rdatatable/data.table#4259]
Currently for non-UTC `POSIXct` input, `as.IDate` is doing:

```
as.IDate(as.Date(x, tz = if (is.null(tz)) "" else tz, ...))
```

and in turn `as.Date.POSIXct` is doing:

```
as.Date(as.POSIXlt(x, tz = tz))
```

More direct to just use `as.IDate.POSIXlt` if we have it.

This method is very easy to write, basically

```
with(x, as.Date(sprintf('%04d-%02d-%02d', year, mon - 1L, mday)))
```

**Comment by aadler** [comment]
If @s-u could be persuaded to release https://github.com/s-u/fasttime under the MPL, that would be amazing. If not, is there a way to clean room design something which performs a similar function?

**Comment by MichaelChirico** [comment]
as in, getting around `POSIXlt` altogether?

conditional on having parsed the input as `POSIXlt`, I think an efficient approach is possible, just need to be careful about leap years/centuries.

**Comment by markseeto** [comment]
@MichaelChirico I've been thinking about this. While `as.IDate.POSIXlt` could be written, I don't know if it would be useful in defining `as.IDate.POSIXct`, because `as.IDate` has different behaviour for POSIXlt input compared to POSIXct input.

```
library(data.table)

x <- as.POSIXlt("2023-09-23 05:00", tz = "Australia/Sydney")

as.IDate(x, tz = "UTC")
# [1] "2023-09-23"

as.IDate(as.POSIXct(x), tz = "UTC")
# [1] "2023-09-22"
```
I assume this is intentional, because `as.Date` does the same thing.

```
as.Date(x, tz = "UTC")
# [1] "2023-09-23"

as.Date(as.POSIXct(x), tz = "UTC")
# [1] "2023-09-22"
```
If I'm misunderstanding something, please let me know.

**Comment by atrayees** [comment]
@MichaelChirico will this piece of code be sufficient for the as.IDate.POSIXlt method? I just wanted to clarify before attempting a PR. 

```
as.IDate.POSIXlt = function(x, tz = attr(x, "tzone", exact = TRUE), ...) {
  if (is_utc(tz))
    (setattr(as.integer(as.numeric(x) %/% 86400L), "class", c("IDate", "Date")))
  else
    as.IDate(with(x, as.Date(sprintf('%04d-%02d-%02d', year, mon - 1L, mday))), tz = tz, ...)
}
```

As for the suggested version, what I'm understanding is that one way to do it could be creating a helper function with the year, month and day as parameters; this function will then do the calculations for the leap years/centuries. Hope I am correct?

**Comment by tdhock** [comment]
I don't know, I'm not familiar with this code.

**Comment by corsairier** [comment]
Hey @MichaelChirico, Is the issue still relevant? If so could I work on this issue?

**Comment by MichaelChirico** [comment]
Sure, I think it makes sense to have a go at it. Be aware that I'm not 100% sure this method is needed, but it is worth pursuing to the point of a PR.

**Comment by Mukulyadav2004** [comment]
Hi @MichaelChirico 

I've tried implementing the solution to add an `as.IDate.POSIXlt` method and use it for `as.IDate.POSIXct` as per your suggestion.

My implementation aimed to correctly handle the behavioral difference between `POSIXlt` and `POSIXct` inputs that @markseeto pointed out. Correctness tests showed the output was identical to the current method. 

However, a performance benchmark on a large vector revealed an unexpected result: the new method was significantly slower for non-UTC dates.
My guess is that the R-level call to `as.POSIXlt()` may be creating significant overhead. 

Does this performance result suggest this approach isn't viable?

Please let me know how you'd like me to proceed.

**Comment by MichaelChirico** [comment]
Thanks, let's close this, it looks not worth the effort.

If you could share the code you wrote, that would help for if anyone decides to revisit this eventually.

**Comment by Mukulyadav2004** [comment]
Certainly, and thank you for the guidance. Here are the R code changes I used for testing, followed by the benchmark script that demonstrated the performance issue.

In `R/IDateTime.R`
```
as.IDate.POSIXlt = function(x, ...) {
  as.IDate(sprintf('%04d-%02d-%02d', x$year + 1900L, x$mon + 1L, x$mday))
}

as.IDate.POSIXct = function(x, tz = attr(x, "tzone", exact=TRUE), ...) {
  if (is_utc(tz)) {
    (setattr(as.integer(as.numeric(x) %/% 86400L), "class", c("IDate", "Date")))
  } else {
    as.IDate(as.POSIXlt(x, tz = tz %||% ''))
  }
}
```

Benchmarking script I used for testing:
```
library(microbenchmark)
 
x = as.POSIXct("2025-06-12", tz="Asia/Kolkata") + 1:1e6
 
microbenchmark(
  "Current" = as.IDate(as.Date(x, tz = "Asia/Kolkata")),
  "After changes" = as.IDate(x, tz = "Asia/Kolkata"),
  times = 10L
)
```

**Comment by s-u** [comment]
@Mukulyadav2004 conversions between `POSIXlt` and `POSIXct` are very expensive since they are system calls which have to deal with TZ and DST for each datum separately.

Also note that R's `POSIXlt` method for `as.Date` does NOT have the `tz` argument, which is why you are seeing the different results (and why `data.table:::as.IDate.default` currently does the same of ignoring `tz` for `POSIXlt`). Since `IDate` is supposed to mimic `Date` it should be ok to not use `tz` either.

PS: I only now saw the comment above about `fasttime` license (you'll have more success by e-mailing me) - I don't care I'm happy to make it MIT... (but `fasttime` only parses UTC which is why it is fast).

**Comment by Mukulyadav2004** [comment]
Thank you very much @s-u, for the expert insight. 
Your explanation that these conversions trigger individual system calls perfectly clarifies the performance regression I observed in my benchmark. It's a much deeper reason than I had initially guessed.

I also appreciate the confirmation regarding the design of `as.Date.POSIXlt` and why it intentionally lacks a `tz` argument.

**Comment by s-u** [comment]
Lol, not sure about the "intentionally" part ... may have been either expedience or historical. It is not explicitly mentioned in the help page so a discussion on R-devel would be good to either document it or discuss the options ...

**Comment by Mukulyadav2004** [comment]
Thanks for the additional context. If a discussion on R-devel ever comes up, I'll be very interested to follow it.
